voxblox/include/voxblox/integrator/tsdf_integrator.h

```diff
@@ -155,109 +316,93 @@ class TsdfIntegratorBase {
       Block<TsdfVoxel>::Ptr block;
-
-      for (const AnyIndex& global_voxel_idx : global_voxel_indices) {
-        BlockIndex block_idx = getBlockIndexFromGlobalVoxelIndex(
-            global_voxel_idx, voxels_per_side_inv_);
-        VoxelIndex local_voxel_idx =
-            getLocalFromGlobalVoxelIndex(global_voxel_idx, voxels_per_side_);
-
-        if (local_voxel_idx.x() < 0) {
-          local_voxel_idx.x() += voxels_per_side_;
-        }
-        if (local_voxel_idx.y() < 0) {
-          local_voxel_idx.y() += voxels_per_side_;
-        }
-        if (local_voxel_idx.z() < 0) {
-          local_voxel_idx.z() += voxels_per_side_;
-        }
-
-        if (!block || block_idx != last_block_idx) {
-          block = layer_->allocateBlockPtrByIndex(block_idx);
-          block->updated() = true;
-          last_block_idx = block_idx;
+      VoxelIndex global_voxel_idx;
+      while (ray_caster.nextRayIndex(&global_voxel_idx)) {
+        TsdfVoxel* voxel = getVoxel(global_voxel_idx, &block, &last_block_idx);
+        if (voxel != nullptr) {
+          const float weight = getVoxelWeight(point_C);
+          const Point voxel_center_G =
+              getCenterPointFromGridIndex(global_voxel_idx, voxel_size_);
+
+          updateTsdfVoxel(origin, point_G, voxel_center_G, color,
+                          config_.default_truncation_distance, weight, voxel);
         }
-
-        const Point voxel_center_G =
-            block->computeCoordinatesFromVoxelIndex(local_voxel_idx);
-        TsdfVoxel& tsdf_voxel = block->getVoxelByVoxelIndex(local_voxel_idx);
-
-        const float weight = getVoxelWeight(point_C);
-        updateTsdfVoxel(origin, point_G, voxel_center_G, color,
-                        truncation_distance, weight, &tsdf_voxel);
       }
-      update_voxels_timer.Stop();
     }
-    integrate_timer.Stop();
   }
 };
 
 class MergedTsdfIntegrator : public TsdfIntegratorBase {
  public:
   MergedTsdfIntegrator(const Config& config, Layer<TsdfVoxel>* layer)
-      : TsdfIntegratorBase(config, layer){};
+      : TsdfIntegratorBase(config, layer) {}
 
   void integratePointCloud(const Transformation& T_G_C,
                            const Pointcloud& points_C, const Colors& colors) {
-    DCHECK_EQ(points_C.size(), colors.size());
+    timing::Timer block_timer("block_allocation");
+    allocateBlocks(T_G_C, points_C, colors);
+    block_timer.Stop();
+
     timing::Timer integrate_timer("integrate");
 
+    point_getter_ =
+        std::make_shared<PointGetter>(T_G_C, points_C, colors, config_);
+
     // Pre-compute a list of unique voxels to end on.
     // Create a hashmap: VOXEL INDEX -> index in original cloud.
     BlockHashMapType<std::vector<size_t>>::type voxel_map;
     // This is a hash map (same as above) to all the indices that need to be
     // cleared.
     BlockHashMapType<std::vector<size_t>>::type clear_map;
 
-    bundleRays(T_G_C, points_C, &voxel_map, &clear_map);
+    bundleRays(&voxel_map, &clear_map);
```

Comment:
While I have no opinion on whether to refactor PointGetter or not, I agree with Marius on the name obfuscating its purpose. Consider changing it to something that implies it also "stores" the data...